### ISSUE TITLE
Can't use TUS endpoint with a subfolder

### ISSUE THREAD

**Issue by mlumbroso** [tus/tus-js-client#148]
Hello,

First thanks for this amazing library !

I've posted my issue here: https://github.com/transloadit/uppy/issues/1432. But after checking, it seems it's probably more linked to TUS than Uppy. 
My issue is I can't use the TUS upload on a subfolder endpoint.
I'm using on the server tus-node-server version 0.3.2, configured as per the documentation :
```
const uploadServer = new tus.Server();
uploadServer.datastore = new tus.FileStore({
    path: '/uploads'
});
const uploadApp = express();
uploadApp.all('*', uploadServer.handle.bind(uploadServer));
app.use('/upload', uploadApp);
```

On the front-end I'm using TUS with uppy like this :

```
uppy.use(Uppy.Tus, {
   endpoint: 'https://example.com/subfolder/upload/', // use your tus endpoint here
   resume: true,
   autoRetry: true,
   retryDelays: [0, 1000, 3000, 5000]
});
```

If I include the subfolder there, it breaks 2 things :
1) The HEAD call in the beginning is made to *https://example.com/upload/uploads/fileID* instead of *https://example.com/subfolder/upload/uploads/fileID* resulting in a 404 error

2) The PATCH request at the end also returns a 404 since it's hitting the same wrong URL.

Please note that the upload and the upload progress are working fine since they are hitting the correct endpoint.
If I remove the subfolder from endpoint it's working perfectly !

From my first checkings on the code, I think the issue comes from file `lib/upload.js` where for the upload we use `this.endpoint`  directly.

Whereas for the HEAD and PATCH requests we use `this.url`, which is declared as is :
`this.url = resolveUrl(this.options.endpoint, location);`

Let me know if you need any additional details or if you could point me to the file where I could create a PR for this. 
Thanks :)

PS : it seems like Uppy is not using the latest version of tus-js-client (https://github.com/ifedapoolarewaju/tus-js-client#888bcf73b66698a165f086f7bbe61951597f5c1b ), not sure if the issue comes from there or if it's still present on master.

**Comment by Acconut** [comment]
Thanks for the detailed description. However, this does not sound like an issue with tus-js-client but rather that your server is misconfigured. Where does the `subfolder` in your endpoint come from? Your tus server is configured to accept uploads at `/uploads/upload` and not `/subfolder/uploaders/upload`.  Are you using a proxy in between which rewrites URLs? I am having problems understanding what you are trying to achieve.

**Comment by mlumbroso** [comment]
@Acconut thanks for your answer.
Sorry if this wasn't very clear, my server is deployed on a subfolder of my main domain. My nginx configuration is as followed :
```   
location /subfolder {
         .... // stripped the useless info, just to show the proxy
            proxy_pass http://127.0.0.1:3270;
    }
```
So requests are made to https://example.com/subfolder/upload/uploads

That is working great for the uploads as said in the first post, the server is configured correctly and I can see the upload progress, since from what I saw on the code the upload is hitting the "raw endpoint" as is : `xhr.open("POST", this.options.endpoint, true);` (https://github.com/tus/tus-js-client/blob/master/lib/upload.js#L300 )

The issue is with the HEAD and PATCH calls which are using `this.url` (https://github.com/tus/tus-js-client/blob/master/lib/upload.js#L360 and https://github.com/tus/tus-js-client/blob/master/lib/upload.js#L448 )

Inside the this.url, it's the endpoint but stripped of the subfolder, which is why it is triggering the 404 error in my case. That's why I figured there might be an issue with the declaration here: https://github.com/tus/tus-js-client/blob/master/lib/upload.js#L314 

Hope it is clearer now, thanks for your help :)

**Comment by mlumbroso** [comment]
@Acconut I investigated and realized the issue comes from the Location header sent from the Node server.
I thought I should first try to fix it from tus-node-server side, but I couldn't figure out a solution to get the full URL of the request from the server.

I sort of worked out a (really not clean) solution from the client that is working for me, but I don't think it has any chance to be merged since I'm quite not sure of the impact it can have and there should be a better way of doing it.
If you want to take a look at what I changed, my commit is here : https://github.com/mlumbroso/tus-js-client/commit/58d0abbe5fb5fe0268ab24239d2020f897fcd785

Maybe you will have a better idea than me on how to fix it. (not sure if the issue happens also with tusd tbh). Please let me know your thoughts :)

**Comment by Acconut** [comment]
> I investigated and realized the issue comes from the Location header sent from the Node server.
I thought I should first try to fix it from tus-node-server side, but I couldn't figure out a solution to get the full URL of the request from the server.

Yes, and that's also the expected behavior with your setup. The problem is that the proxy rewrites the URL (i.e. adds a subfolder) without the tus-node-server knowing about this. Therefore the tus-node-server returns the "incorrect" Location header because it doesn't know about the proxy's URL rewriting. On the client side, tus-js-client relies on tus-node-server to deliver the correct upload URL, which is unfortunately incorrect.

So, the problem's source does not lay inside tus-js-client and also not inside tus-node-server because the proxy is kind of messing around with both sides. As far as I know, tus-node-server does not have an option to handle subfolders added by proxies directly but there should be another solution.

> I sort of worked out a (really not clean) solution from the client that is working for me, but I don't think it has any chance to be merged since I'm quite not sure of the impact it can have and there should be a better way of doing it.

Yes, I am sorry but in its current form there is little chance that this is going to be merged :)

> Maybe you will have a better idea than me on how to fix it. (not sure if the issue happens also with tusd tbh). Please let me know your thoughts :)

(Yes, I assume it would also happen with tusd). Since the proxy is the source of the problem, I think the best solution is to solve it using the proxy. Basically, what you need to do is to also rewrite the Location header in the POST response to also include the subfolder (i.e. from `/upload/uploads/...` to `/subfolder/upload/uploads...`). In that case, tus-js-client will obtain the correct upload URL and everything should work out.

**Comment by mlumbroso** [comment]
Thanks for the detailed answer, makes sense :)

I think I'll need to use `proxy_redirect` in addition to `proxy_pass`.
Will test it this week and will share in tus-node-server if I'm successful about it, so maybe we can add a line on the README about this !

**Comment by mlumbroso** [comment]
I've performed some tests.
The only way that I was able to make it work is to change the Host header like this :
`proxy_set_header Host $host/subfolder;`
instead of `proxy_set_header Host $host;`

Cause on tus-node-server it builds the Location header like this :
```const url = this.store.relativeLocation ? `${req.baseUrl || ''}${this.store.path}/${File.id}` : `//${req.headers.host}${req.baseUrl || ''}${this.store.path}/${File.id}`;```

Maybe I missed something cause I'm not sure it's really standard to put a subfolder on the Host header, but at least that fixes it. @Acconut or @bhstahl if you have any thoughts if this is the correct way of fixing my issue :) 

Thanks for the help !

**Comment by Acconut** [comment]
Adding a path to the Host headers sounds dangerous to me. Did you attempt to rewrite the Location header instead? If you enable `relativeLocation` in tus-node-server and then let the proxy add `/subfolder` in front of the Location headers in the *POST response*, it should also work and be a bit saner :)

**Comment by dmytrohridin** [comment]
hey @Acconut 
I have a similar issue, but unfortunately I can not change server implementation. Maybe it make sense to make ```resolveUrl``` function to be configurable with default implementation.
Thank you

**Comment by Acconut** [comment]
@dmytrohridin Please open a new issue with your details.

**Comment by hrvoj3e** [comment]
I will just leave this here because this issue is ranked high on google when searching for problems with subfolder deploy of tus server. 

https://gist.github.com/hrvoj3e/77cb8e63ba126de17a8e31cb5d750da4

TL;DR: use `proxy_redirect ~/files/(.*)$ $scheme://$host/my-subfolder/files/$1;`
